boa_engine/src/builtins/temporal/calendar/object.rs

```diff
@@ -0,0 +1,299 @@
+//! Boa's implementation of a user-defined Anonymous Calendar.
+
+use crate::{
+    builtins::temporal::{plain_date, plain_date_time, plain_month_day, plain_year_month},
+    object::ObjectKind,
+    property::PropertyKey,
+    Context, JsObject, JsValue,
+};
+
+use boa_macros::utf16;
+use boa_temporal::{
+    calendar::{CalendarDateLike, CalendarProtocol},
+    date::TemporalDate,
+    duration::Duration,
+    error::TemporalError,
+    fields::TemporalFields,
+    month_day::TemporalMonthDay,
+    options::ArithmeticOverflow,
+    year_month::TemporalYearMonth,
+    TemporalResult, TinyStr4, TinyStr8,
+};
+
+/// A user-defined, custom calendar that is only known at runtime
+/// and executed at runtime.
+///
+/// A user-defined calendar implements all methods of the CalendarProtocol,
+/// and therefore satisfies the requirements to be used as a calendar.
+#[derive(Debug, Clone)]
+pub(crate) struct CustomRuntimeCalendar {
+    calendar: JsObject,
+    ctx: *mut Context,
+}
+
+impl CustomRuntimeCalendar {
+    pub(crate) fn new(calendar: &JsObject, context: &mut Context) -> Self {
+        Self {
+            calendar: calendar.clone(),
+            ctx: context,
+        }
+    }
+}
+
+impl CalendarProtocol for CustomRuntimeCalendar {
+    fn date_from_fields(
+        &self,
+        fields: &mut TemporalFields,
+        overflow: ArithmeticOverflow,
+    ) -> TemporalResult<TemporalDate> {
+        // Safety: Context lives for the life of the program and execution, so
+        // this should, in theory, be valid.
+        let context = unsafe { &mut *self.ctx };
```

Comment:
I don't think this is valid. Any `JsObject` could live indefinitely right now, even after its containing context was dropped. Maybe in the future we could support it with the GC redesign, but at the moment we'll need to change our approach.

That being said, I think we can receive a `&mut dyn Any` parameter on the trait and downcast to `&mut Context` inside the implementation.

Reply:
Looks like that should actually work perfectly. 😄